File: 23.链表中环的入口结点/23.链表中环的入口结点.py
# -*- coding:utf-8 -*-
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
class Solution:
    def EntryNodeOfLoop(self, pHead):
        # write code here
        slow = pHead
        fast = pHead
        while(fast is not None and fast.next is not None):
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                fast = pHead
                while(fast!=slow):
                    fast = fast.next
                    slow = slow.next
                return slow
        return None

File: 23.链表中环的入口结点/23.链表中环的入口结点.cpp
/*
struct ListNode {
    int val;
    struct ListNode *next;
    ListNode(int x) :
        val(x), next(NULL) {
    }
};
*/
class Solution {
public:
    ListNode* EntryNodeOfLoop(ListNode* pHead)
    {
        ListNode *fast = pHead,*slow = pHead;
        while(fast!=nullptr && fast->next != nullptr){
        	slow = slow->next;
            fast = fast->next->next;
            if(fast == slow){
				fast = pHead;
                while(fast!=slow){
                    fast = fast->next;
                    slow = slow->next;
                }
            }
            if(fast==slow){
                return fast;
            }
        
        }
        return nullptr;
    }
        
};